### DISCUSSION TITLE
Urlaubsmodus/Partymodus

### DISCUSSION THREAD

**Discussion by Mephisto090** [OpenCCU/OpenCCU#2901] [🇩🇪 Fragen&Antworten] [answered]
### Describe the issue you are experiencing

Hallo,

ich habe nun seit mehreren Stunden versucht, mithilfe von Skripten und Systemvariablen einen Urlaubs- bzw. Partymodus zu aktivieren, jedoch leider ohne Erfolg. Weder die Einstellung von SET_POINT_MODE auf 2 noch die Aktivierung von PARTY_MODE auf true führen zum gewünschten Ergebnis.

Obwohl ich es geschafft habe, dass mein HmIP anzeigt, wann der Urlaub endet, erfolgt der Wechsel in den entsprechenden Modus nicht automatisch. Auch die Einstellung von PARTY_TIME_START scheint nicht korrekt zu funktionieren (siehe Bild).

Ich würde mich über Unterstützung freuen, um dieses Problem zu lösen.

Vielen Dank!

### Describe the behavior you expected

Start via Skript den Urlaubs bzw. Partymodus

### Steps to reproduce the issue

Siehe wieter oben

### What is the version this bug report is based on?

3.79.6.20241031

### Which base platform are you running?

rpi4 (RaspberryPi4, ARM64/aarch64)

### Which HomeMatic/homematicIP radio module are you using?

RPI-RF-MOD

### Anything in the logs that might be useful for us?

```txt
...
```

### Additional information

![image](https://github.com/user-attachments/assets/a90c2c17-011b-478e-8858-ad652341b5f4)
![image](https://github.com/user-attachments/assets/137dcb30-737f-4695-ae07-098162132d05)
![image](https://github.com/user-attachments/assets/1df34faa-39f5-4ae9-8e0e-88ede8495798)

```
! ID des HmIP-Geräts, das du steuern möchtest
string deviceId = "HmIP-RF.00395XXXXXXX";  ! Ersetze <DEINE-GERÄTE-ID> durch die Geräte-ID des Thermostats

! Ziel-Solltemperatur für den Urlaubsmodus einstellen
real targetTemp = dom.GetObject("Urlaub_Temperatur").Value();  ! Beispiel-Solltemperatur in °C

! Start- und Endzeit des Urlaubsmodus festlegen
string startDate = dom.GetObject("Urlaub_Start").Value();  ! Startdatum und -uhrzeit im Format JJJJ_MM_TT HH:MM
string endDate = dom.GetObject("Urlaub_Ende").Value();    ! Enddatum und -uhrzeit im Format JJJJ_MM_TT HH:MM

! Solltemperatur im Urlaubsmodus einstellen
! dom.GetObject(deviceId + ":1.PARTY_SET_POINT_TEMPERATURE").State(targetTemp);

! Urlaubsmodus aktivieren und Start-/Endzeit festlegen
! dom.GetObject(deviceId + ":1.PARTY_MODE").State(true);

dom.GetObject(deviceId + ":1.PARTY_TIME_START").State(startDate);
dom.GetObject(deviceId + ":1.PARTY_TIME_END").State(endDate);
dom.GetObject(deviceId + ":1.SET_POINT_MODE").State(2);
dom.GetObject(deviceId + ":1.SET_POINT_TEMPERATURE").State(targetTemp);
```

**Comment by jens-maus** [comment]
> Ich würde mich über Unterstützung freuen, um dieses Problem zu lösen.

Das ist ein Bug/Issue-Tracker hier und kein Supportforum. Bitte an einschlägige Diskussionsforen für Hilfestellungen wenden und diesen Bug/Issue-Tracker nur für bestätigte Bugs benutzen die auf ein Problem in RaspberryMatic selbst hinweisen.

**Reply by Mephisto090** [reply]
Naja, da die Variablen anscheinend nicht richtig gesetzt wurden, habe ich angenommen, dass es sich um einen Bug handelt oder dass einige Variablen nicht ordnungsgemäß funktionieren.

Was mache ich als Benutzer also falsch? Ich bin offen für Unterstützung und würde gerne wissen, wie ich den Urlaubsmodus über ein Skript aktivieren kann. Vielen Dank im Voraus für eure Hilfe!

**Reply by jens-maus** [reply]
Na ich sehe schon bei dem...

```
dom.GetObject(deviceId + ":1.PARTY_TIME_START").State(startDate);
```

Das du da offensichtlich nen Fehler hast weil du da fälschlicherweise `+` statt `#` für eine string concatenation nutzt...

**Reply by Mephisto090** [reply]
Alles klar, habe ich durch den folgenen Quelltext ersetzt. Leider wird er immer noch nicht gestartet. Desweiteren wird die Variable `hm-rpc.0.00395F29B25XXX.1.PARTY_TIME_START` auch nicht richtig gesetzt. Wo könnte da der Fehler liegen?

![image](https://github.com/user-attachments/assets/225f8e8f-b6ba-4039-99dc-1e94b4eed9e1)

```
! ID des HmIP-Geräts, das du steuern möchtest
string deviceId = "HmIP-RF.00395F29B25XXX";  ! Ersetze <DEINE-GERÄTE-ID> durch die Geräte-ID des Thermostats

! Ziel-Solltemperatur für den Urlaubsmodus einstellen
real targetTemp = dom.GetObject("Urlaub_Temperatur").Value();  ! Beispiel-Solltemperatur in °C

! Start- und Endzeit des Urlaubsmodus festlegen
string startDate = dom.GetObject("Urlaub_Start").Value();  ! Startdatum und -uhrzeit im Format JJJJ_MM_TT HH:MM
string endDate = dom.GetObject("Urlaub_Ende").Value();    ! Enddatum und -uhrzeit im Format JJJJ_MM_TT HH:MM

! Solltemperatur im Urlaubsmodus einstellen
dom.GetObject(deviceId # ":1.PARTY_SET_POINT_TEMPERATURE").State(targetTemp);

! Urlaubsmodus aktivieren und Start-/Endzeit festlegen
! dom.GetObject(deviceId # ":1.PARTY_MODE").State(true);

dom.GetObject(deviceId # ":1.PARTY_TIME_START").State(startDate);
dom.GetObject(deviceId # ":1.PARTY_TIME_END").State(endDate);
dom.GetObject(deviceId # ":1.SET_POINT_MODE").State(2);
! dom.GetObject(deviceId # ":1.SET_POINT_TEMPERATURE").State(targetTemp);
```

**Reply by jens-maus** [reply]
> Desweiteren wird die Variable `hm-rpc.0.00395F29B25XXX.1.PARTY_TIME_START` auch nicht richtig gesetzt.

Na wenn du schon in der ioBroker-Welt bist, warum nutzt du dann nicht die Programmierumgebung (javascript, blockly, usw.) von ioBroker und versuchst das in der Homematic-Skript Welt zu erledigen?!?

**Reply by Mephisto090** [reply]
Auch das habe ich versucht. Vars wurden nicht richtig von ioBroker an CCU übertragen. Das "gleich" hatte ich dort schon aufgebaut und versucht

**Reply by Mephisto090** [reply]
Bekomme das im iobroker angezeigt;
![image](https://github.com/user-attachments/assets/63ea47f9-5a63-4edb-9c3c-bc98fcde5ae8)

Ändere es in das hier:
![image](https://github.com/user-attachments/assets/8507fc9b-bcb1-4e15-93c4-1915ba230860)

5 Sekunden später habe ich von der ccu wieder das hier im ioBroker:
![image](https://github.com/user-attachments/assets/dfc972cf-8d43-44bc-be4e-7b80341d340b)

**Reply by Mephisto090** [reply]
Schau mal was die RaspberryMatic hier komisches macht. Uhrzeit war 20:51 am 04.11.2024

![Animation](https://github.com/user-attachments/assets/2222e331-5272-4920-9778-85d1a26ba2ce)

![image](https://github.com/user-attachments/assets/e7fb010a-091c-41b5-809e-ad146ef5d970)

**Reply by Mephisto090** [reply]
@jens-maus nach etwas testen funktioniert zwar, wenn ich bei der CCU den Urlaubsmodus per Hand starte das er `hm-rpc.0.00395F29B25XXX.1.SET_POINT_MODE` auf `2` setzt aber nicht von ioBroker zur CCU. Dieser Wert wird patu nicht übernommen. ggf. doch ein Bug oder was mache ich falsch?

**Comment by Baxxy13** [comment]
Siehe: https://homematic-forum.de/forum/viewtopic.php?f=26&t=70560&hilit=urlaubsmodus+script&start=10#p687009
und folgende Posts.

**Reply by Mephisto090** [reply]
Super, das hat schon mal funktioniert! 

Dazu hätte ich noch eine Fragen:

Gibt es eine Möglichkeit, alle bzw. mehrere HmIP-Geräte gleichzeitig anzusteuern?

Vielen Dank vorab für deine Unterstützung!

**Comment by Mephisto090** [comment, marked as answer]
```
! - PutParamset Zuweisung für mehrere Geräte im Urlaubsmodus
string stdErr = "";
string stdOut = "";
string TCL = "";

! Start- und Endzeit des Urlaubsmodus festlegen
string deviceList = "Zimmer:1,Raum:1,Bad:1, ....";  ! Liste der Geräte, getrennt durch Kommas
string startDate = dom.GetObject("start_var").Value();  ! Startdatum und -uhrzeit im Format JJJJ_MM_TT HH:MM
string endDate = dom.GetObject("ende_var").Value();    ! Enddatum und -uhrzeit im Format JJJJ_MM_TT HH:MM
real targetTemp = dom.GetObject("temp_var").Value();  ! Beispiel-Solltemperatur in °C

! TCL-Befehle für Urlaubmodus erstellen
TCL = TCL # " {SET_POINT_MODE {i4 2}}";
TCL = TCL # " {SET_POINT_TEMPERATURE {double " # targetTemp # "}}";
TCL = TCL # " {PARTY_TIME_START {" # startDate # "}}";
TCL = TCL # " {PARTY_TIME_END {" # endDate # "}}";

! Schleife über alle Geräte in der Liste
integer i = 0;
string device = deviceList.StrValueByIndex(",", i);

while (device != "") {
    object oCHANNEL = dom.GetObject(ID_CHANNELS).Get(device);
    if (oCHANNEL) {  ! Überprüfen, ob das Gerät existiert
        object oIFace = dom.GetObject(oCHANNEL.Interface());
        string sTCL = ^puts [xmlrpc ^ #oIFace.InterfaceUrl()# ^/ putParamset [list string \"^#oCHANNEL.Address()#^\"] [list string \"VALUES\"] [list struct \"^#TCL#^\"]]^;
        system.Exec("/bin/sh -c 'echo \"load tclrpc.so; " # sTCL # "\" |tclsh'", &stdOut, &stdErr);
    }
    i = i + 1;
    device = deviceList.StrValueByIndex(",", i);  ! Nächstes Gerät abrufen
}

! Optional: Ausgabe der Ergebnisse
WriteLine("Urlaubsmodus für alle Geräte gesetzt. Fehlerausgabe: " # stdErr);

```

**Comment by Baxxy13** [comment]
Na geht doch. 😉 
Ich hoffe der Duty Cycle spielt mit wenn du das auf deine (unbestimmte Anzahl an) Thermostate(n) loslässt.

**Reply by Mephisto090** [reply]
Sind nur 8 und da geht es. Da ist er bei rund 40%

**Reply by jens-maus** [reply]
Trotzdem sieht das sehr waghalsig aus mit tclsh Aufruf via System.exec()...

**Reply by Mephisto090** [reply]
Hast du einen anderen Vorschlag, wie man es besser machen könnte?

**Reply by jens-maus** [reply]
Geht es denn nicht direkt via ReGa Skriptfunktion `xmlrpc.PutParamset()` (siehe https://github.com/jens-maus/RaspberryMatic/issues/1769) statt dieses "Von hinten durch die Brust ins Auge"? oder vielleicht via ioBroker Funktionalitäten, siehe https://github.com/ioBroker/ioBroker.hm-rpc#custom-commands

**Reply by Baxxy13** [reply]
Müsste eigentlich gehen.
Die Schwierigkeit dabei ist, das nicht nur ein Parameter samt Wert übergeben werden soll, sondern jeweils 4 in einem Rutsch.
Soweit ich mich erinnere habe ich das damals nicht hinbekommen und habe mich deshalb der "Blaupause" aus dem SDV bedient.
😉

**Reply by jens-maus** [reply]
Wenn es dafür Änderungen an ReGa bräuchte... Ich bin ja da ;-)

**Reply by Baxxy13** [reply]
Könnte auch wieder nur ne Escaping-Geschichte sein...
Bzw. ich weiß nicht genau wie das Array/Struct mit den Parametern/Werten "aussehen" muss.

Aktuell sieht der Befehl über tclrpc.so so aus:
`puts [xmlrpc xmlrpc://127.0.0.1:32010/ putParamset [list string \"000A98A9A228C2:1\"] [list string \"VALUES\"] [list struct \" {SET_POINT_MODE {i4 2}} {SET_POINT_TEMPERATURE {double 13.500000}} {PARTY_TIME_START {2024_01_09 17:00}} {PARTY_TIME_END {2024_01_09 18:30}}\"]]`

**Reply by Baxxy13** [reply]
Ein einzelner Parameter ist einfach...
`xmlrpc.PutParamset(1009,000A98A9A228C2:1,VALUES,SET_POINT_MODE,16.000000);`

Aber wie packe ich noch 3 weitere Parameter samt Wert dazu?

**Reply by jens-maus** [reply]
Ich befürchte aktuell kann die `xmlrpc.Putparamset()` Funktion nur jeweils einen Parameter (z.B. `SET_POINT_MODE`) und einen wert dazu (hier `16.00000`) nehmen und intern verarbeiten. Da müsste ich mir erst noch überlegen wie man der Funktion abwärtskompatibel beibringen kann auch mehrere Parameter und Werte auf einmal anzunehmen und zu verarbeiten.

**Reply by Baxxy13** [reply]
Ah, das erklärt warum ich das nie hinbekommen habe. 
Das Issue was du verlinkt hast ist  ja auch noch offen, vielleicht kannst du da ja auch was machen wenn du an die Methode rangehst.

**Reply by Mephisto090** [reply]
Ich hatte es auch versucht und deswegen jetzt die "schöne" Lösung, die super funktioniert.
